pkg/filter/manager.go

```diff
@@ -0,0 +1,276 @@
+	case api.StreamFilterStop:
+		return StreamFilterChainReset
+	case api.StreamFiltertermination:
+		return StreamFilterChainReset
+	}
+
+	return StreamFilterChainContinue
+}
+
+// StreamFilterManager manages the lifecycle of streamFilters.
+type StreamFilterManager interface {
+	// register StreamSenderFilter, StreamReceiverFilter and AccessLog.
+	api.StreamFilterChainFactoryCallbacks
+
+	// invoke the receiver filter chain.
+	RunReceiverFilter(ctx context.Context, phase api.FilterPhase,
+		headers types.HeaderMap, data types.IoBuffer, trailers types.HeaderMap,
+		statusHandler StreamFilterStatusHandler) api.StreamFilterStatus
+
+	// invoke the sender filter chain.
+	RunSenderFilter(ctx context.Context, phase api.FilterPhase,
+		headers types.HeaderMap, data types.IoBuffer, trailers types.HeaderMap,
+		statusHandler StreamFilterStatusHandler) api.StreamFilterStatus
+
+	// invoke all access log.
+	Log(ctx context.Context, reqHeaders api.HeaderMap, respHeaders api.HeaderMap, requestInfo api.RequestInfo)
+
+	// destroy the sender filter chain and receiver filter chain.
+	OnDestroy()
+}
+
+// StreamReceiverFilterWithPhase combines the StreamReceiverFilter with its Phase.
+type StreamReceiverFilterWithPhase interface {
+	api.StreamReceiverFilter
+	ValidatePhase(phase api.FilterPhase) bool
+}
+
+// StreamReceiverFilterWithPhaseImpl is the default implementation of StreamReceiverFilterWithPhase.
+type StreamReceiverFilterWithPhaseImpl struct {
+	filter api.StreamReceiverFilter
+	phase  api.FilterPhase
+}
+
+// NewStreamReceiverFilterWithPhaseImpl returns a StreamReceiverFilterWithPhaseImpl struct..
+func NewStreamReceiverFilterWithPhaseImpl(
+	f api.StreamReceiverFilter, p api.FilterPhase) *StreamReceiverFilterWithPhaseImpl {
+	return &StreamReceiverFilterWithPhaseImpl{
+		filter: f,
+		phase:  p,
+	}
+}
+
+// OnDestroy destroys the StreamReceiverFilter.
+func (s *StreamReceiverFilterWithPhaseImpl) OnDestroy() {
+	s.filter.OnDestroy()
+}
+
+// OnReceive invokes the StreamReceiverFilter.
+func (s *StreamReceiverFilterWithPhaseImpl) OnReceive(ctx context.Context,
+	headers api.HeaderMap, buf buffer.IoBuffer, trailers api.HeaderMap) api.StreamFilterStatus {
+	return s.filter.OnReceive(ctx, headers, buf, trailers)
+}
+
+// SetReceiveFilterHandler sets the filterHandler for StreamReceiverFilter.
+func (s *StreamReceiverFilterWithPhaseImpl) SetReceiveFilterHandler(handler api.StreamReceiverFilterHandler) {
+	s.filter.SetReceiveFilterHandler(handler)
+}
+
+// ValidatePhase checks the current phase.
+func (s *StreamReceiverFilterWithPhaseImpl) ValidatePhase(phase api.FilterPhase) bool {
+	return s.phase == phase
+}
+
+// StreamSenderFilterWithPhase combines the StreamSenderFilter which its Phase.
+type StreamSenderFilterWithPhase interface {
+	api.StreamSenderFilter
+	ValidatePhase(phase api.FilterPhase) bool
+}
+
+// StreamSenderFilterWithPhaseImpl is default implementation of StreamSenderFilterWithPhase.
+type StreamSenderFilterWithPhaseImpl struct {
```

Comment:
Can be optimized in the same way as `StreamReceiverFilterWithPhaseImpl` comments.